Unit tests for extract_text, harden up WARC parsing

davidmcclure opened the pull request:
- Harden up the [file type detection and text extraction](https://github.com/opensyllabus/osp-pipeline/blob/extract-text-tests/osp_pipeline/jobs/extract_text/scraper_warc.py#L41) logic in `ScraperWARC`.
- Add [factories](https://github.com/opensyllabus/osp-pipeline/blob/extract-text-tests/tests/factories/warc.py) (via FactoryBoy) to programmatically mock WARCs produced by the scraper.
- [Unit test](https://github.com/opensyllabus/osp-pipeline/tree/extract-text-tests/tests/osp_pipeline/jobs/extract_text/scraper_warc) `ScraperWARC`.
- Break out a [`Text.from_scraper_warc()`](https://github.com/opensyllabus/osp-pipeline/blob/extract-text-tests/osp_pipeline/schemas/text.py#L21) factory method, which maps a WARC blob into a Text row, unit test this.
- Under `tests/`, nest the actual test files under an `osp_pipeline` directory, to mirror the application code. This is just so that we can put testing helpers under `tests/` without worrying about name conflicts with application modules (eg, if we wanted to have a `utils` module for testing helpers, but also want to test a `utils` module in the application code.)
- Get rid of old `corpus` tests, which are folded in to the `ScraperWARC` tests.

Sorry this is a big diff, one thing led to another :)

As for testing the interactions with s3 - I tried a couple of things here (moto, fakes3, pytest mocks), all of which had problems. Moto breaks pyspark when it tries to write output. Fakes3 involves all kinds of changes to the application code to get it to point at localhost instead of s3, though it might be worth trying to get this to work, since it seems nice in concept. I also tried directly patching the `read_bytes` method that pulls data from s3, but this doesn't work when run under Spark, because the executors re-import the code and never see the patched method.

It works to copy fixtures to live s3, but this is also complex and slow. @wearpants, let me know if you have ideas for other approaches to this.

wearpants commented:
> I also tried directly patching the read_bytes method that pulls data from s3, but this doesn't work when run under Spark, because the executors re-import the code and never see the patched method.

The solution here would be to do the patching using `tests.hooks` (from the patches), which is the first thing executed in child Python processes

davidmcclure commented:
A change in `test.hooks` would be a global for the entire suite though, right? What we'd want is a way to do it programmatically for individual tests. Eg, in test X, `read_bytes` should return WARCs A, B, and C, with specific fields, etc.

wearpants commented:
This is using mocks? No, won't work then, thought it was for a different
strategy.

On Thu, Mar 2, 2017 at 4:09 PM, David McClure <notifications@github.com>
wrote:

> A change in test.hooks would be a global for the entire suite though,
> right? What we'd want is a way to do it programmatically for individual
> tests. Eg, in test X, read_bytes should return WARCs A, B, and C, with
> specific fields, etc.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/opensyllabus/osp-pipeline/pull/24#issuecomment-283780759>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAe2ACTKOBLJAfD6PSudhCfXrforr1Vkks5rhy_ygaJpZM4MQoPz>
> .
>



-- 
Pete Fein | wearpants.org | @wearpants

wearpants commented:
Thinking about this more, you could  use the hooks to patch, and return
specific content based on filename. You'd need to define the
filename->content mapping at the point of patching rather than with a mock
in the individual test itself. This is trivial if your data comes from an
actual  file (my original tests did this), a little more involved if using
a factory (create warcs in RAM when patching).  Make sense?

On Mar 2, 2017 4:20 PM, "Peter Fein" <pete@wearpants.org> wrote:

> This is using mocks? No, won't work then, thought it was for a different
> strategy.
>
> On Thu, Mar 2, 2017 at 4:09 PM, David McClure <notifications@github.com>
> wrote:
>
>> A change in test.hooks would be a global for the entire suite though,
>> right? What we'd want is a way to do it programmatically for individual
>> tests. Eg, in test X, read_bytes should return WARCs A, B, and C, with
>> specific fields, etc.
>>
>> —
>> You are receiving this because your review was requested.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/opensyllabus/osp-pipeline/pull/24#issuecomment-283780759>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAe2ACTKOBLJAfD6PSudhCfXrforr1Vkks5rhy_ygaJpZM4MQoPz>
>> .
>>
>
>
>
> --
> Pete Fein | wearpants.org | @wearpants
>

wearpants commented on osp_pipeline/jobs/extract_text/scraper_warc.py:
```diff
@@ -13,6 +14,12 @@
 )
 
 
+class FileType(Enum):
```
I like everything that is happening here. :-)

wearpants commented on tests/conftest.py:
```diff
@@ -9,30 +9,37 @@
         .getOrCreate()
     )
 
-    request.addfinalizer(lambda: sc.stop())
```
Questions about removing pytest finalizer, see above

davidmcclure replied:
Refactored this a bit in 01337b5ac7b1e740262185ece59165590f54d84e. Before, I had `spark` as a `yield_fixture` that pulled in `sc` and called `sc.stop()` after the yield, which means it would always run at the end of the suite. But, since this was called in `spark`, and not `sc`, `stop()` wouldn't get called in the (unlikely) case that the suite uses `sc` but never `spark`.

But, either way, it makes sense to do the cleanup on `sc` in the `sc` fixture. So now we've got:

```python
@pytest.yield_fixture(scope='session')
def sc():
    """Provide a SparkContext.
    """
    conf = (
        SparkConf()
        .setMaster('local[4]')
        .setAppName('pytest-pyspark-local-testing')
    )

    conf.set('spark.python.worker.reuse', 'true')

    sc = SparkContext(conf=conf)

    yield sc

    sc.stop()
```

Which guarantees that `stop()` always runs at the end of the suite.

wearpants commented on tests/conftest.py:
```diff
@@ -9,30 +9,37 @@
 @pytest.fixture(scope='session')
-def spark(request):
-    """Provide a spark SparkSession.
+def sc():
```
I am concerned by the removal of the `request` arg here, and the `finalizer` below. I don't remember why I did it this way, but the spark/pytest was generally very temperamental, and I'd be inclined to revert it unless you had a reason not to.

wearpants commented on tests/factories/warc.py:
```diff
@@ -0,0 +1,195 @@
+class WARCFactory(factory.Factory):
+
+    class Meta:
+        model = io.BytesIO
```
Well this is clever :-)

davidmcclure replied:
Hey, why not, right? 😄

davidmcclure commented on tests/conftest.py:
```diff
@@ -9,30 +9,37 @@
 
     The SparkContext is available as `spark.sparkContext`.
 
     :arg request: pytest.FixtureRequest object
```
Fixed by c09c328a7172d273467e955f0658315bfb71e2de.

davidmcclure commented on tests/conftest.py:
```diff
@@ -9,30 +9,37 @@
+
+    The SparkContext is available as `spark.sparkContext`.
+
+    :arg request: pytest.FixtureRequest object
```
Fixed by c09c328a7172d273467e955f0658315bfb71e2de.

davidmcclure commented:
Thanks for this @wearpants - I fixed the items flagged on the code comments (see responses inline), and updated the README about writing tests and Invoke.

The docs for writing jobs still look good to me?

davidmcclure commented:
@wearpants, is this good to go?

wearpants commented:
Looks good

On Mar 9, 2017 8:40 PM, "David McClure" <notifications@github.com> wrote:

> @wearpants <https://github.com/wearpants>, is this good to go?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opensyllabus/osp-pipeline/pull/24#issuecomment-285547697>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAe2AGt5tLMRVjVmL33kG03jZgbU9eyyks5rkKoFgaJpZM4MQoPz>
> .
>
